### DISCUSSION TITLE
attitude PID

### DISCUSSION THREAD

**Discussion by fangjir** [bitcraze/discussions#1398] [General]
Today, when I was looking at the underlying program, I found that the attitude loop PID is in the form of PI+PID. Theoretically speaking, if the attitude loop is divided into inner and outer loops to control, PID cannot be used.
![微信图片_20240613171800](https://github.com/bitcraze/discussions/assets/71303114/eb02e634-6609-4f7e-8e40-6129c202d24f)
![微信图片_20240613171806](https://github.com/bitcraze/discussions/assets/71303114/a2582141-00ee-4413-a200-b343be67f2b2)

**Comment by knmcguire** [comment]
Hi!  The PID controller is of a cascaded form: https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/controllers/#cascaded-pid-controller

So if the setpoint for roll/pitch/yaw is indicated to be in velocity mode (aka ratemode), then the other attitude controller is not used but just the rate attitude controller is. Indeed, in the attitude range, the D-gain isn't used much, but it has been tuned by hand so perhaps some improvements can be made.

Did that answer your question? If not, would you mind formulating it more clearly?

**Reply by fangjir** [reply]
@knmcguire  thanks,I see what you mean. However, I still want to know whether the D term is used here because the output is directly PWM, and the moment and PWM can be equivalent to a first-order inertia link, so the attitude loop model has become a third-order system, so D can be used.
